// @posthog/wizard/dist/src/steps/add-mcp-server-to-clients/clients/visual-studio-code.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VisualStudioCodeClient = exports.VisualStudioCodeMCPConfig = void 0;
const zod_1 = __importDefault(require("zod"));
const path = __importStar(require("path"));
const os = __importStar(require("os"));
const MCPClient_1 = require("../MCPClient");
exports.VisualStudioCodeMCPConfig = zod_1.default
    .object({
    servers: zod_1.default.record(zod_1.default.string(), zod_1.default.object({
        command: zod_1.default.string().optional(),
        args: zod_1.default.array(zod_1.default.string()).optional(),
        env: zod_1.default.record(zod_1.default.string(), zod_1.default.string()).optional(),
    })),
})
    .passthrough();
class VisualStudioCodeClient extends MCPClient_1.DefaultMCPClient {
    name = 'Visual Studio Code';
    getServerPropertyName() {
        return 'servers';
    }
    async isClientSupported() {
        return Promise.resolve(process.platform === 'darwin' ||
            process.platform === 'win32' ||
            process.platform === 'linux');
    }
    async getConfigPath() {
        const homeDir = os.homedir();
        const isWindows = process.platform === 'win32';
        const isMac = process.platform === 'darwin';
        const isLinux = process.platform === 'linux';
        if (isMac) {
            return Promise.resolve(path.join(homeDir, 'Library', 'Application Support', 'Code', 'User', 'mcp.json'));
        }
        if (isWindows) {
            return Promise.resolve(path.join(process.env.APPDATA || '', 'Code', 'User', 'mcp.json'));
        }
        if (isLinux) {
            return Promise.resolve(path.join(homeDir, '.config', 'Code', 'User', 'mcp.json'));
        }
        throw new Error(`Unsupported platform: ${process.platform}`);
    }
}
exports.VisualStudioCodeClient = VisualStudioCodeClient;
