// @actbase/native/lib/devtools/common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handleCopy = exports.isClipboardEnabled = exports.NETWORK_LOG_BACKGROUND = exports.METHOD_OPTIONS = exports.READY_STATUS = exports.MINUS_SIZE = exports.BUBBLE_SIZE = void 0;

var _reactNative = require("react-native");

var BUBBLE_SIZE = 58;
exports.BUBBLE_SIZE = BUBBLE_SIZE;
var MINUS_SIZE = -4;
exports.MINUS_SIZE = MINUS_SIZE;
var READY_STATUS = {
  OPENED: 1,
  HEADERS_RECEIVED: 2,
  LOADING: 3,
  DONE: 4
};
exports.READY_STATUS = READY_STATUS;
var METHOD_OPTIONS = {
  GET: {
    label: 'GET',
    background: '#61affe'
  },
  POST: {
    label: 'POST',
    background: '#49cc90'
  },
  PUT: {
    label: 'PUT',
    background: '#fca130'
  },
  DELETE: {
    label: 'DEL',
    background: '#f93e3e'
  }
};
exports.METHOD_OPTIONS = METHOD_OPTIONS;
var NETWORK_LOG_BACKGROUND = {
  '4': '#ffffe4',
  '5': '#ffe4e4'
};
exports.NETWORK_LOG_BACKGROUND = NETWORK_LOG_BACKGROUND;
var nativeClipboard = _reactNative.NativeModules.RNCClipboard;
var isClipboardEnabled = !!nativeClipboard;
exports.isClipboardEnabled = isClipboardEnabled;

var handleCopy = function handleCopy(str) {
  if (!nativeClipboard) return;

  try {
    nativeClipboard.setString(str);

    _reactNative.Alert.alert('Actbase', 'Copy to Data');
  } catch (e) {
    _reactNative.Alert.alert('Actbase', 'Copy Failure');
  }
};

exports.handleCopy = handleCopy;
