// @actbase/native/lib/devtools/console.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _styles = _interopRequireDefault(require("./styles"));

var _common = require("./common");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Console = function Console(_a) {
  var data = _a.data,
      onClear = _a.onClear;
  return _react["default"].createElement(_reactNative.FlatList, {
    data: data,
    contentContainerStyle: {
      flexGrow: 1
    },
    ListEmptyComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].noItem
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].noItemText
    }, "Not found Console Log")),
    ListHeaderComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].header
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].headerTitle
    }, "Console"), !!(data === null || data === void 0 ? void 0 : data.length) && _react["default"].createElement(_reactNative.TouchableOpacity, {
      style: _styles["default"].clearButton,
      onPress: onClear
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].clearButtonText
    }, "CLEAR"))),
    renderItem: function renderItem(_a) {
      var _b;

      var item = _a.item;
      var dt = new Date();
      dt.setTime(item.time);
      var timeStr = [];
      timeStr.push(dt.getHours().toString().padStart(2, '0'));
      timeStr.push(dt.getMinutes().toString().padStart(2, '0'));
      timeStr.push(dt.getSeconds().toString().padStart(2, '0'));
      return _react["default"].createElement(_reactNative.View, {
        style: _styles["default"].consoleItem
      }, _react["default"].createElement(_reactNative.View, {
        style: {
          flexDirection: 'row',
          justifyContent: 'space-between',
          alignItems: 'center'
        }
      }, _react["default"].createElement(_reactNative.View, {
        style: _styles["default"].consoleItemTag
      }, _react["default"].createElement(_reactNative.Text, {
        style: _styles["default"].consoleItemTagText
      }, (_b = item.type) === null || _b === void 0 ? void 0 : _b.toUpperCase())), _react["default"].createElement(_reactNative.View, {
        style: {
          flexDirection: 'row',
          alignItems: 'center'
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 10,
          color: '#555'
        }
      }, timeStr === null || timeStr === void 0 ? void 0 : timeStr.join(':')), _common.isClipboardEnabled && _react["default"].createElement(_reactNative.TouchableOpacity, {
        style: {
          marginLeft: 5
        },
        onPress: function onPress() {
          return (0, _common.handleCopy)(item.body);
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12
        }
      }, "\uD83D\uDCC4")))), _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12,
          color: '#555',
          lineHeight: 16
        }
      }, item.body));
    },
    style: {
      flex: 1
    }
  });
};

var _default = Console;
exports["default"] = _default;
