// @actbase/native/lib/devtools/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.useDevTools = exports.DevContext = void 0;

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

var _tools = _interopRequireDefault(require("./tools"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var INIT_OPTIONS = {
  module: {
    console: true,
    network: true
  }
};

var DevContext = _react["default"].createContext({
  options: INIT_OPTIONS
});

exports.DevContext = DevContext;

var useDevTools = function useDevTools() {
  var o = _react["default"].useContext(DevContext);

  return {
    showTools: o.showTools,
    setNavigation: o.setNavigation,
    setVariants: o.setVariants
  };
};

exports.useDevTools = useDevTools;

var DevTools = function DevTools(_a) {
  var debug = _a.debug,
      children = _a.children,
      reduxStore = _a.reduxStore;

  var _b = (0, _react.useState)(!debug),
      disabled = _b[0],
      setDisabled = _b[1];

  var _c = (0, _react.useState)(INIT_OPTIONS),
      options = _c[0],
      setOptions = _c[1];

  var _d = (0, _react.useState)(reduxStore),
      redux = _d[0],
      setRedux = _d[1];

  var _e = (0, _react.useState)({}),
      variants = _e[0],
      _setVariants = _e[1];

  var _f = (0, _react.useState)(undefined),
      navigation = _f[0],
      setNavigation = _f[1];

  (0, _react.useEffect)(function () {
    if (debug && !(typeof debug === 'boolean')) {
      setOptions(function (x) {
        var _a, _b, _c, _d, _e, _f;

        return (0, _tslib.__assign)((0, _tslib.__assign)({}, x), {
          module: (0, _tslib.__assign)((0, _tslib.__assign)({}, x.module), {
            console: (_c = (_a = debug === null || debug === void 0 ? void 0 : debug.console) !== null && _a !== void 0 ? _a : (_b = x.module) === null || _b === void 0 ? void 0 : _b.console) !== null && _c !== void 0 ? _c : false,
            network: (_f = (_d = debug === null || debug === void 0 ? void 0 : debug.network) !== null && _d !== void 0 ? _d : (_e = x.module) === null || _e === void 0 ? void 0 : _e.network) !== null && _f !== void 0 ? _f : false
          })
        });
      });
    }

    return undefined;
  }, [debug]);
  (0, _react.useEffect)(function () {
    setRedux(reduxStore);
  }, [reduxStore]);

  var _g = (0, _react.useMemo)(function () {
    var e = disabled ? _react.Fragment : _tools["default"];
    return [e, !disabled ? (0, _tslib.__assign)((0, _tslib.__assign)({}, options), {
      redux: redux,
      navigation: navigation,
      variants: variants
    }) : {}];
  }, [disabled, options, navigation, variants]),
      ToolElement = _g[0],
      toolProps = _g[1];

  return _react["default"].createElement(DevContext.Provider, {
    value: {
      showTools: function showTools() {
        return setDisabled(false);
      },
      options: options,
      setOptions: setOptions,
      setNavigation: setNavigation,
      setVariants: function setVariants(o) {
        _setVariants(function (x) {
          return (0, _tslib.__assign)((0, _tslib.__assign)({}, x), o);
        });
      }
    }
  }, _react["default"].createElement(ToolElement, (0, _tslib.__assign)({}, toolProps), children));
};

var _default = _react["default"].memo(DevTools);

exports["default"] = _default;
