// @actbase/native/lib/devtools/navigation.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = _interopRequireDefault(require("./styles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var parseState = function parseState(data) {
  var _a, _b, _c;

  if (!(data === null || data === void 0 ? void 0 : data.routes) || !(data === null || data === void 0 ? void 0 : data.routeNames) || !((_a = data === null || data === void 0 ? void 0 : data.routes) === null || _a === void 0 ? void 0 : _a.forEach)) return [];

  try {
    var o_1 = (_c = (_b = data === null || data === void 0 ? void 0 : data.routeNames) === null || _b === void 0 ? void 0 : _b.map(function (name) {
      return {
        name: name
      };
    })) !== null && _c !== void 0 ? _c : [];
    data.routes.forEach(function (item) {
      var ix = o_1.findIndex(function (v) {
        return v.name === item.name;
      });

      if (ix >= 0) {
        o_1[ix].key = item.key;
        o_1[ix].items = parseState(item.state);
        o_1[ix].data = JSON.stringify(o_1[ix].items);
      }
    });
    return o_1;
  } catch (e) {
    return [];
  }
};

var NavigationItem = function NavigationItem(_a) {
  var _b, _c;

  var item = _a.item,
      current = _a.current,
      navigate = _a.navigate;
  return _react["default"].createElement(_reactNative.View, null, _react["default"].createElement(_reactNative.TouchableOpacity, {
    style: _styles["default"].navigationItem,
    activeOpacity: 0.8,
    onPress: function onPress() {
      return navigate(item.name);
    }
  }, _react["default"].createElement(_reactNative.Text, {
    style: _styles["default"].navigationText
  }, item.name), current === item.key && _react["default"].createElement(_reactNative.View, {
    style: {
      width: 5,
      height: 5,
      borderRadius: 2.5,
      backgroundColor: '#db2929'
    }
  })), !!((_b = item === null || item === void 0 ? void 0 : item.items) === null || _b === void 0 ? void 0 : _b.length) && _react["default"].createElement(_reactNative.View, {
    style: _styles["default"].navigationChild
  }, (_c = item === null || item === void 0 ? void 0 : item.items) === null || _c === void 0 ? void 0 : _c.map(function (o, index) {
    return _react["default"].createElement(NavigationItem, {
      item: o,
      key: String(index),
      current: current,
      navigate: navigate
    });
  })));
};

var Navigation = function Navigation(_a) {
  var navigation = _a.navigation,
      opened = _a.opened,
      onExit = _a.onExit;

  var _b = (0, _react.useState)([]),
      data = _b[0],
      setData = _b[1];

  var _c = (0, _react.useState)(''),
      current = _c[0],
      setCurrent = _c[1];

  (0, _react.useEffect)(function () {
    var _a, _b, _c, _d, _e, _f;

    if (!(navigation === null || navigation === void 0 ? void 0 : navigation.current) || !opened) return;
    var state = (_b = (_a = navigation.current) === null || _a === void 0 ? void 0 : _a.getRootState) === null || _b === void 0 ? void 0 : _b.call(_a);
    var key = ((_e = (_d = (_c = navigation.current) === null || _c === void 0 ? void 0 : _c.getCurrentRoute) === null || _d === void 0 ? void 0 : _d.call(_c)) !== null && _e !== void 0 ? _e : {}).key;
    setCurrent(key !== null && key !== void 0 ? key : '');
    setData((_f = parseState(state)) !== null && _f !== void 0 ? _f : []);
  }, [opened]);

  var navigate = function navigate(name) {
    var _a;

    if (!((_a = navigation === null || navigation === void 0 ? void 0 : navigation.current) === null || _a === void 0 ? void 0 : _a.navigate)) return;
    navigation.current.navigate(name);
    if (onExit) onExit();
  };

  return _react["default"].createElement(_reactNative.ScrollView, {
    style: {
      flex: 1
    }
  }, _react["default"].createElement(_reactNative.View, {
    style: _styles["default"].header
  }, _react["default"].createElement(_reactNative.Text, {
    style: _styles["default"].headerTitle
  }, "Navigation")), data === null || data === void 0 ? void 0 : data.map(function (item, index) {
    var _a;

    return _react["default"].createElement(NavigationItem, {
      item: item,
      key: (_a = item.key) !== null && _a !== void 0 ? _a : String(index),
      current: current,
      navigate: navigate
    });
  }));
};

var _default = Navigation;
exports["default"] = _default;
