// @actbase/native/lib/devtools/network_badge.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _common = require("./common");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var BACKGROUND_FOR_STATE = {
  ok: '#339933',
  fail: '#db2929',
  ready: '#aaa'
};

var NetworkBadge = function NetworkBadge(_a) {
  var _b, _c, _d;

  var data = _a.data,
      show = _a.show;
  var opacity = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
  var tk = (0, _react.useRef)(0);
  var items = (0, _react.useMemo)(function () {
    return data.slice(0, 3);
  }, [data]);
  var first = items === null || items === void 0 ? void 0 : items[0];
  (0, _react.useEffect)(function () {
    if (tk.current && typeof tk.current === 'number') {
      clearTimeout(tk.current);
    }

    _reactNative.Animated.timing(opacity, {
      toValue: (data === null || data === void 0 ? void 0 : data.length) > 0 ? 1 : 0,
      duration: 150,
      useNativeDriver: true
    }).start();

    var f = data === null || data === void 0 ? void 0 : data[0];

    if ((data === null || data === void 0 ? void 0 : data.length) > 0 && (f === null || f === void 0 ? void 0 : f.state) !== 'ready') {
      tk.current = setTimeout(function () {
        _reactNative.Animated.timing(opacity, {
          toValue: 0,
          duration: 150,
          useNativeDriver: true
        }).start();
      }, 2000);
    }
  }, [((_b = data === null || data === void 0 ? void 0 : data[0]) === null || _b === void 0 ? void 0 : _b.state) + "-" + ((_c = data === null || data === void 0 ? void 0 : data[0]) === null || _c === void 0 ? void 0 : _c.time)]);
  if (!show) return null;
  var methodOption = _common.METHOD_OPTIONS[first === null || first === void 0 ? void 0 : first.method];
  return _react["default"].createElement(_reactNative.Animated.View, {
    style: [_reactNative.StyleSheet.absoluteFill, {
      opacity: opacity,
      backgroundColor: '#fff',
      alignItems: 'center',
      justifyContent: 'center'
    }]
  }, _react["default"].createElement(_reactNative.View, {
    style: {
      flexDirection: 'row',
      marginTop: 4
    }
  }, items.map(function (_a) {
    var state = _a.state,
        url = _a.url,
        time = _a.time;
    return !state ? undefined : _react["default"].createElement(_reactNative.View, {
      key: time + "-" + url,
      style: {
        backgroundColor: BACKGROUND_FOR_STATE === null || BACKGROUND_FOR_STATE === void 0 ? void 0 : BACKGROUND_FOR_STATE[state],
        width: 4,
        height: 4,
        borderRadius: 2,
        marginHorizontal: 1
      }
    });
  })), methodOption !== undefined && _react["default"].createElement(_reactNative.View, {
    style: {
      backgroundColor: methodOption.background,
      borderRadius: 2,
      paddingHorizontal: 4,
      height: 15,
      alignItems: 'center',
      justifyContent: 'center',
      marginTop: 3
    }
  }, _react["default"].createElement(_reactNative.Text, {
    style: {
      fontSize: 10,
      fontWeight: 'bold',
      color: '#fff'
    },
    allowFontScaling: false
  }, methodOption.label)), _react["default"].createElement(_reactNative.Text, {
    style: {
      fontSize: 8,
      lineHeight: 12,
      height: 12
    },
    allowFontScaling: false
  }, !(first === null || first === void 0 ? void 0 : first.finish) ? '-' : (first === null || first === void 0 ? void 0 : first.finish) - ((_d = first === null || first === void 0 ? void 0 : first.time) !== null && _d !== void 0 ? _d : 0)));
};

var _default = NetworkBadge;
exports["default"] = _default;
