// @actbase/native/lib/devtools/network.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _common = require("./common");

var _styles = _interopRequireDefault(require("./styles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var FIELDS = [{
  key: 'url',
  title: 'URL'
}, {
  key: 'status',
  title: 'Status Code'
}, {
  key: 'method',
  title: 'Method'
}, {
  key: 'headers',
  title: 'Headers'
}, {
  key: 'body',
  title: 'Request Body'
}, {
  key: 'response',
  title: 'Response Text'
}];

var Network = function Network(_a) {
  var data = _a.data,
      onClear = _a.onClear;

  var _b = (0, _react.useState)(),
      details = _b[0],
      setDetails = _b[1];

  return _react["default"].createElement(_reactNative.View, {
    style: {
      flex: 1
    }
  }, details && _react["default"].createElement(_reactNative.View, {
    style: [_reactNative.StyleSheet.absoluteFill, {
      zIndex: 10,
      backgroundColor: '#fff',
      borderRadius: 20
    }]
  }, _react["default"].createElement(_reactNative.View, null, _react["default"].createElement(_reactNative.TouchableOpacity, {
    onPress: function onPress() {
      return setDetails(undefined);
    },
    style: {
      height: 40,
      justifyContent: 'center',
      paddingHorizontal: 10
    }
  }, _react["default"].createElement(_reactNative.Text, {
    style: {
      fontSize: 12
    }
  }, "Go Back"))), _react["default"].createElement(_reactNative.ScrollView, {
    style: {
      flex: 1
    }
  }, FIELDS === null || FIELDS === void 0 ? void 0 : FIELDS.map(function (field) {
    // eslint-disable-next-line no-nested-ternary
    var text = typeof (details === null || details === void 0 ? void 0 : details[field.key]) === 'string' ? details === null || details === void 0 ? void 0 : details[field.key] : JSON.stringify(details === null || details === void 0 ? void 0 : details[field.key], null, 2);
    if (!text) return null;
    return _react["default"].createElement(_reactNative.View, {
      key: field.key,
      style: {
        paddingHorizontal: 10,
        paddingBottom: 15
      }
    }, _react["default"].createElement(_reactNative.View, {
      style: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center'
      }
    }, _react["default"].createElement(_reactNative.Text, {
      style: {
        fontSize: 11,
        color: '#333',
        fontWeight: 'bold',
        lineHeight: 16
      }
    }, field.title), _common.isClipboardEnabled && _react["default"].createElement(_reactNative.TouchableOpacity, {
      onPress: function onPress() {
        return (0, _common.handleCopy)(text);
      }
    }, _react["default"].createElement(_reactNative.Text, {
      style: {
        fontSize: 12
      }
    }, "\uD83D\uDCC4"))), _react["default"].createElement(_reactNative.TouchableOpacity, {
      onPress: function onPress() {
        return (0, _common.handleCopy)(text);
      },
      disabled: !_common.isClipboardEnabled,
      style: {
        backgroundColor: '#eee',
        borderRadius: 5,
        padding: 10,
        marginTop: 5
      }
    }, _react["default"].createElement(_reactNative.Text, {
      style: {
        fontSize: 12,
        color: '#555'
      }
    }, text)));
  }))), _react["default"].createElement(_reactNative.FlatList, {
    data: data,
    contentContainerStyle: {
      flexGrow: 1
    },
    keyExtractor: function keyExtractor(_, index) {
      return "" + index;
    },
    ListEmptyComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].noItem
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].noItemText
    }, "Not found Network Log")),
    ListHeaderComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].header
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].headerTitle
    }, "Network"), !!(data === null || data === void 0 ? void 0 : data.length) && _react["default"].createElement(_reactNative.TouchableOpacity, {
      style: _styles["default"].clearButton,
      onPress: onClear
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].clearButtonText
    }, "CLEAR"))),
    renderItem: function renderItem(_a) {
      var _b, _c, _d, _e, _f;

      var item = _a.item;
      var opt = _common.METHOD_OPTIONS[item.method];
      var date = new Date();

      if (item.time) {
        date.setTime(item.time);
      }

      var timeStr = [];
      timeStr.push(date.getHours().toString().padStart(2, '0'));
      timeStr.push(date.getMinutes().toString().padStart(2, '0'));
      timeStr.push(date.getSeconds().toString().padStart(2, '0'));
      var e = String(Math.floor(((_b = item.status) !== null && _b !== void 0 ? _b : 0) / 100));
      return _react["default"].createElement(_reactNative.TouchableOpacity, {
        onPress: function onPress() {
          delete item.obj;
          setDetails(item);
        },
        style: [_styles["default"].networkItem, {
          backgroundColor: (_c = _common.NETWORK_LOG_BACKGROUND === null || _common.NETWORK_LOG_BACKGROUND === void 0 ? void 0 : _common.NETWORK_LOG_BACKGROUND[e]) !== null && _c !== void 0 ? _c : '#fff'
        }]
      }, _react["default"].createElement(_reactNative.View, {
        style: [_styles["default"].networkItemTag, {
          backgroundColor: opt === null || opt === void 0 ? void 0 : opt.background
        }]
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 10,
          fontWeight: 'bold',
          color: '#fff'
        },
        allowFontScaling: false
      }, item.method), _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 10,
          fontWeight: 'bold',
          color: '#fff'
        }
      }, (_d = item.status) !== null && _d !== void 0 ? _d : '-')), _react["default"].createElement(_reactNative.View, {
        style: {
          flex: 1
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12,
          color: '#333'
        }
      }, item.url), _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 10,
          lineHeight: 12,
          marginTop: 2,
          color: '#777'
        }
      }, item.finish && item.time && _react["default"].createElement(_react["default"].Fragment, null, ((_e = item.finish) !== null && _e !== void 0 ? _e : 0) - ((_f = item.time) !== null && _f !== void 0 ? _f : 0), "ms", ' - '), "Call to ", timeStr === null || timeStr === void 0 ? void 0 : timeStr.join(':'))));
    },
    style: {
      flex: 1
    }
  }));
};

var _default = Network;
exports["default"] = _default;
