// @actbase/native/lib/devtools/redux.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = _interopRequireDefault(require("./styles"));

var _common = require("./common");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Redux = function Redux(_a) {
  var store = _a.store,
      opened = _a.opened;

  var _b = (0, _react.useState)([]),
      data = _b[0],
      setData = _b[1];

  var _c = (0, _react.useState)([]),
      openeds = _c[0],
      setOpeneds = _c[1];

  (0, _react.useEffect)(function () {
    if (!store || !opened) return;
    var v = store.getState();
    setData(Object.keys(v || {}).reduce(function (a, key) {
      var rowData = {
        key: key,
        value: JSON.stringify(v[key], null, 2)
      };
      a.push(rowData);
      return a;
    }, []));
  }, [store, opened]);
  return _react["default"].createElement(_reactNative.FlatList, {
    data: data,
    contentContainerStyle: {
      flexGrow: 1
    },
    ListEmptyComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].noItem
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].noItemText
    }, "Not found Data")),
    ListHeaderComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].header
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].headerTitle
    }, "Redux Store")),
    renderItem: function renderItem(_a) {
      var item = _a.item;
      var open = openeds.includes(item.key);
      return _react["default"].createElement(_reactNative.View, null, _react["default"].createElement(_reactNative.TouchableOpacity, {
        activeOpacity: 0.9,
        style: {
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          paddingHorizontal: 10,
          minHeight: 35,
          paddingVertical: 5,
          borderBottomWidth: 1,
          borderBottomColor: '#eee'
        },
        onPress: function onPress() {
          setOpeneds(function (x) {
            var draft = (0, _tslib.__spreadArrays)(x);
            var ix = draft.indexOf(item.key);

            if (ix >= 0) {
              draft.splice(ix, 1);
            } else {
              draft.push(item.key);
            }

            return draft;
          });
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12,
          fontWeight: 'bold',
          color: '#555'
        }
      }, item.key), _react["default"].createElement(_reactNative.View, {
        style: {
          flexDirection: 'row',
          alignItems: 'center'
        }
      }, _react["default"].createElement(_reactNative.View, {
        style: {
          width: 6,
          height: 6,
          borderLeftWidth: 2,
          borderBottomWidth: 2,
          right: 3,
          top: open ? 1 : -1,
          transform: [{
            rotateZ: open ? '135deg' : '-45deg'
          }]
        }
      }), _common.isClipboardEnabled && _react["default"].createElement(_reactNative.TouchableOpacity, {
        style: {
          marginLeft: 5
        },
        onPress: function onPress() {
          return (0, _common.handleCopy)(item.value);
        },
        hitSlop: {
          left: 10,
          right: 10,
          bottom: 10,
          top: 10
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12
        }
      }, "\uD83D\uDCC4")))), open && _react["default"].createElement(_reactNative.View, {
        style: {
          backgroundColor: '#f0f0f0',
          paddingHorizontal: 10,
          paddingTop: 10,
          paddingBottom: 20
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12
        }
      }, item.value)));
    },
    style: {
      flex: 1
    }
  });
};

var _default = Redux;
exports["default"] = _default;
