// @actbase/native/lib/devtools/storage.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.isAsyncStorage = void 0;

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = _interopRequireDefault(require("./styles"));

var _common = require("./common");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var RCTAsyncStorage = _reactNative.NativeModules.PlatformLocalStorage || // Support for external modules, like react-native-windows
_reactNative.NativeModules.RNC_AsyncSQLiteDBStorage || _reactNative.NativeModules.RNCAsyncStorage;

var isAsyncStorage = function isAsyncStorage() {
  return !!RCTAsyncStorage;
};

exports.isAsyncStorage = isAsyncStorage;

var AsyncStorage = function AsyncStorage(_a) {
  var opened = _a.opened;

  var _b = (0, _react.useState)(),
      err = _b[0],
      setErr = _b[1];

  var _c = (0, _react.useState)([]),
      data = _c[0],
      setData = _c[1];

  var _d = (0, _react.useState)([]),
      openeds = _d[0],
      setOpeneds = _d[1];

  (0, _react.useEffect)(function () {
    if (!opened) return;
    RCTAsyncStorage.getAllKeys(function (error, keys) {
      if (error) {
        setErr(error.message);
        return;
      }

      RCTAsyncStorage.multiGet(keys, function (errors, result) {
        if ((errors === null || errors === void 0 ? void 0 : errors.length) > 0) {
          setErr(errors === null || errors === void 0 ? void 0 : errors.map(function (v) {
            return v.message;
          }).join('\n'));
          return;
        }

        setData(result === null || result === void 0 ? void 0 : result.reduce(function (a, v) {
          a.push({
            key: v[0],
            value: v[1]
          });
          return a;
        }, []));
      });
    });
  }, [opened]);
  return _react["default"].createElement(_reactNative.FlatList, {
    data: data,
    contentContainerStyle: {
      flexGrow: 1
    },
    ListEmptyComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].noItem
    }, !err ? _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].noItemText
    }, "Not found Data") : _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].noItemText
    }, err)),
    ListHeaderComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].header
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].headerTitle
    }, "Async Storage")),
    renderItem: function renderItem(_a) {
      var item = _a.item;
      var open = openeds.includes(item.key);
      return _react["default"].createElement(_reactNative.View, null, _react["default"].createElement(_reactNative.TouchableOpacity, {
        activeOpacity: 0.9,
        style: {
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
          paddingHorizontal: 10,
          minHeight: 35,
          paddingVertical: 5,
          borderBottomWidth: 1,
          borderBottomColor: '#eee'
        },
        onPress: function onPress() {
          setOpeneds(function (x) {
            var draft = (0, _tslib.__spreadArrays)(x);
            var ix = draft.indexOf(item.key);

            if (ix >= 0) {
              draft.splice(ix, 1);
            } else {
              draft.push(item.key);
            }

            return draft;
          });
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12,
          fontWeight: 'bold',
          color: '#555'
        }
      }, item.key), _react["default"].createElement(_reactNative.View, {
        style: {
          flexDirection: 'row',
          alignItems: 'center'
        }
      }, _react["default"].createElement(_reactNative.View, {
        style: {
          width: 6,
          height: 6,
          borderLeftWidth: 2,
          borderBottomWidth: 2,
          right: 3,
          top: open ? 1 : -1,
          transform: [{
            rotateZ: open ? '135deg' : '-45deg'
          }]
        }
      }), _common.isClipboardEnabled && _react["default"].createElement(_reactNative.TouchableOpacity, {
        style: {
          marginLeft: 5
        },
        onPress: function onPress() {
          return (0, _common.handleCopy)(item.value);
        },
        hitSlop: {
          left: 10,
          right: 10,
          bottom: 10,
          top: 10
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12
        }
      }, "\uD83D\uDCC4")))), open && _react["default"].createElement(_reactNative.View, {
        style: {
          backgroundColor: '#f0f0f0',
          paddingHorizontal: 10,
          paddingTop: 10,
          paddingBottom: 20
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12
        }
      }, item.value)));
    },
    style: {
      flex: 1
    }
  });
};

var _default = AsyncStorage;
exports["default"] = _default;
