// @actbase/native/lib/devtools/system.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = _interopRequireDefault(require("./styles"));

var _common = require("./common");

var _index = require("./index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var _a, _b, _c;

// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
var constants = (_a = _reactNative.Platform.constants) !== null && _a !== void 0 ? _a : {};
var RNDeviceInfo = _reactNative.NativeModules.RNDeviceInfo;

function formatBytes(bytes, decimals) {
  if (decimals === void 0) {
    decimals = 2;
  }

  if (bytes === 0) return '0 Bytes';
  var k = 1024;
  var dm = decimals < 0 ? 0 : decimals;
  var sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
  var i = Math.floor(Math.log(bytes) / Math.log(k)); // eslint-disable-next-line no-restricted-properties

  return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + " " + sizes[i];
}

var originSections = [{
  key: 'info',
  title: 'System Info',
  data: [{
    title: 'Platform',
    value: _reactNative.Platform.OS
  }, {
    title: 'OS Version',
    value: _reactNative.Platform.Version
  }, constants && {
    title: 'React-Native',
    value: ((_b = constants === null || constants === void 0 ? void 0 : constants.reactNativeVersion) === null || _b === void 0 ? void 0 : _b.major) + "." + ((_c = constants === null || constants === void 0 ? void 0 : constants.reactNativeVersion) === null || _c === void 0 ? void 0 : _c.minor) + "." + (constants === null || constants === void 0 ? void 0 : constants.reactNativeVersion.patch)
  }, RNDeviceInfo && RNDeviceInfo.brand && {
    title: 'Brand',
    value: RNDeviceInfo.brand
  }, RNDeviceInfo && RNDeviceInfo.model && {
    title: 'Model',
    value: RNDeviceInfo.model
  }, RNDeviceInfo && RNDeviceInfo.deviceId && {
    title: 'Device ID',
    value: RNDeviceInfo.deviceId
  }].filter(function (v) {
    return !!v;
  })
}, {
  key: 'var',
  title: 'Variants',
  data: []
}, {
  key: 'cmd',
  title: 'Command',
  data: []
}];

var Switch = function Switch(_a) {
  var value = _a.value;
  return _react["default"].createElement(_reactNative.View, {
    style: {
      width: 6,
      height: 6,
      borderRadius: 3,
      backgroundColor: value ? '#0f0' : '#ccc'
    }
  });
};

var System = function System(_a) {
  var _b, _c;

  var variants = _a.variants;

  var _d = _react["default"].useContext(_index.DevContext),
      options = _d.options,
      setOptions = _d.setOptions;

  var sections = (0, _react.useMemo)(function () {
    var _a, _b, _c, _d, _e, _f;

    var o = JSON.parse(JSON.stringify(originSections));

    try {
      if (_reactNative.Platform.OS === 'android' && RNDeviceInfo && RNDeviceInfo.getDeviceNameSync) {
        o[0].data.push({
          title: 'Device Name',
          value: RNDeviceInfo.getDeviceNameSync()
        });
      }
    } catch (e) {
      o[0].data.push({
        title: 'Device Name',
        value: (_a = e === null || e === void 0 ? void 0 : e.message) !== null && _a !== void 0 ? _a : 'error'
      });
    }

    try {
      if (_reactNative.Platform.OS === 'android' && RNDeviceInfo && RNDeviceInfo.getMacAddressSync) {
        o[0].data.push({
          title: 'MAC Address',
          value: RNDeviceInfo.getMacAddressSync()
        });
      }
    } catch (e) {
      o[0].data.push({
        title: 'MAC Address',
        value: (_b = e === null || e === void 0 ? void 0 : e.message) !== null && _b !== void 0 ? _b : 'error'
      });
    }

    try {
      if (RNDeviceInfo && RNDeviceInfo.getIpAddressSync) {
        o[0].data.push({
          title: 'IP Address',
          value: RNDeviceInfo.getIpAddressSync()
        });
      }
    } catch (e) {
      o[0].data.push({
        title: 'IP Address',
        value: (_c = e === null || e === void 0 ? void 0 : e.message) !== null && _c !== void 0 ? _c : 'error'
      });
    }

    try {
      if (RNDeviceInfo && RNDeviceInfo.getUsedMemorySync) {
        o[0].data.push({
          title: 'Used Memory',
          value: formatBytes(RNDeviceInfo.getUsedMemorySync())
        });
      }
    } catch (e) {
      o[0].data.push({
        title: 'Used Memory',
        value: (_d = e === null || e === void 0 ? void 0 : e.message) !== null && _d !== void 0 ? _d : 'error'
      });
    }

    try {
      if (RNDeviceInfo && RNDeviceInfo.getTotalMemorySync) {
        o[0].data.push({
          title: 'Total Memory',
          value: formatBytes(RNDeviceInfo.getTotalMemorySync())
        });
      }
    } catch (e) {
      o[0].data.push({
        title: 'Total Memory',
        value: (_e = e === null || e === void 0 ? void 0 : e.message) !== null && _e !== void 0 ? _e : 'error'
      });
    }

    try {
      if (RNDeviceInfo && RNDeviceInfo.getFreeDiskStorageSync) {
        o[0].data.push({
          title: 'Free Disk Space',
          value: formatBytes(RNDeviceInfo.getFreeDiskStorageSync())
        });
      }
    } catch (e) {
      o[0].data.push({
        title: 'Free Disk Space',
        value: (_f = e === null || e === void 0 ? void 0 : e.message) !== null && _f !== void 0 ? _f : 'error'
      });
    }

    var keys = Object.keys(variants !== null && variants !== void 0 ? variants : {});

    if (keys.length === 0) {
      o.splice(1, 1);
    } else if (variants) {
      o[1].data = keys.reduce(function (a, b) {
        a.push({
          title: b,
          value: String(variants[b])
        });
        return a;
      }, []);
    }

    var ix = o.findIndex(function (x) {
      return x.key === 'cmd';
    });

    if (_reactNative.NativeModules.DevSettings && __DEV__) {
      o[ix].data.push({
        title: 'Reload',
        onPress: function onPress() {
          _reactNative.NativeModules.DevSettings.reload();
        }
      });
    }

    if (_reactNative.NativeModules.RNExitApp) {
      o[ix].data.push({
        title: 'Exit',
        onPress: function onPress() {
          _reactNative.NativeModules.RNExitApp.exitApp();
        }
      });
    }

    if (o[ix].data.length === 0) {
      o.splice(ix, 1);
    }

    return o;
  }, [variants]);
  return _react["default"].createElement(_reactNative.SectionList, {
    sections: sections,
    contentContainerStyle: {
      flexGrow: 1
    },
    ListHeaderComponent: _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].header
    }, _react["default"].createElement(_reactNative.Text, {
      style: _styles["default"].headerTitle
    }, "System")),
    ListFooterComponent: _react["default"].createElement(_reactNative.View, null, _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].systemSectionHeader
    }, _react["default"].createElement(_reactNative.View, {
      style: _styles["default"].systemSectionBadge
    }, _react["default"].createElement(_reactNative.Text, {
      style: {
        fontSize: 11,
        fontWeight: 'bold',
        color: '#fff'
      }
    }, "Monitor"))), _react["default"].createElement(_reactNative.TouchableOpacity, {
      style: [_styles["default"].systemSectionItem, _styles["default"].systemSectionItemFoot],
      disabled: !setOptions,
      onPress: function onPress() {
        var _a;

        if (!setOptions) return;
        var o = (0, _tslib.__assign)((0, _tslib.__assign)({}, options), {
          module: (0, _tslib.__assign)({}, options.module)
        });
        o.module.network = !((_a = o.module) === null || _a === void 0 ? void 0 : _a.network);
        setOptions(o);
      }
    }, _react["default"].createElement(_reactNative.Text, {
      style: {
        fontSize: 11,
        lineHeight: 16,
        fontWeight: 'bold',
        color: '#333'
      }
    }, "Network"), _react["default"].createElement(Switch, {
      value: !!((_b = options.module) === null || _b === void 0 ? void 0 : _b.network)
    })), _react["default"].createElement(_reactNative.TouchableOpacity, {
      style: [_styles["default"].systemSectionItem, _styles["default"].systemSectionItemFoot],
      disabled: !setOptions,
      onPress: function onPress() {
        var _a;

        if (!setOptions) return;
        var o = (0, _tslib.__assign)((0, _tslib.__assign)({}, options), {
          module: (0, _tslib.__assign)({}, options.module)
        });
        o.module.console = !((_a = o.module) === null || _a === void 0 ? void 0 : _a.console);
        setOptions(o);
      }
    }, _react["default"].createElement(_reactNative.Text, {
      style: {
        fontSize: 11,
        lineHeight: 16,
        fontWeight: 'bold',
        color: '#333'
      }
    }, "Console"), _react["default"].createElement(Switch, {
      value: !!((_c = options.module) === null || _c === void 0 ? void 0 : _c.console)
    }))),
    renderSectionHeader: function renderSectionHeader(_a) {
      var section = _a.section;
      return _react["default"].createElement(_reactNative.View, {
        style: _styles["default"].systemSectionHeader
      }, _react["default"].createElement(_reactNative.View, {
        style: _styles["default"].systemSectionBadge
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 11,
          fontWeight: 'bold',
          color: '#fff'
        }
      }, section.title)));
    },
    renderItem: function renderItem(_a) {
      var _b;

      var item = _a.item;
      return _react["default"].createElement(_reactNative.TouchableOpacity, {
        disabled: !item.onPress,
        onPress: item.onPress,
        style: _styles["default"].systemSectionItem
      }, _react["default"].createElement(_reactNative.View, {
        style: {
          width: 100
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 11,
          lineHeight: 16,
          fontWeight: 'bold',
          color: '#333'
        }
      }, item.title)), item.value !== undefined && _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement(_reactNative.View, {
        style: {
          flex: 1
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12,
          lineHeight: 16,
          color: '#555'
        }
      }, item.value)), _common.isClipboardEnabled && item.value !== undefined && _react["default"].createElement(_reactNative.TouchableOpacity, {
        onPress: function onPress() {
          var _a;

          return (0, _common.handleCopy)((_a = item.value) !== null && _a !== void 0 ? _a : '');
        }
      }, _react["default"].createElement(_reactNative.Text, {
        style: {
          fontSize: 12
        }
      }, "\uD83D\uDCC4"))), item.buttons !== undefined && _react["default"].createElement(_reactNative.View, {
        style: {
          flex: 1,
          justifyContent: 'flex-end',
          flexDirection: 'row'
        }
      }, (_b = item.buttons) === null || _b === void 0 ? void 0 : _b.map(function (button) {
        return _react["default"].createElement(_reactNative.TouchableOpacity, {
          style: {
            marginLeft: 5,
            borderWidth: 1,
            borderColor: '#eee',
            borderRadius: 4,
            paddingHorizontal: 8,
            backgroundColor: '#eee',
            minWidth: 50,
            minHeight: 20,
            alignItems: 'center',
            justifyContent: 'center'
          }
        }, _react["default"].createElement(_reactNative.Text, {
          style: {
            lineHeight: 16,
            fontSize: 11
          }
        }, button.title));
      })));
    },
    style: {
      flex: 1
    }
  });
};

System.defaultProps = {
  variants: {}
};
var _default = System;
exports["default"] = _default;
