// @voiceflow/sdk-runtime/build/esm/runtime/runtime.service.prototype.js
import { RuntimeService } from '../runtime/runtime.service.js';
export class PrototypeRuntimeService extends RuntimeService {
    constructor(options) {
        super(options);
        this.state = {
            turn: {},
            storage: {},
            variables: {},
        };
        this.versionID = options.verify.versionID;
    }
    async interact(request) {
        const { action, config, sessionID } = request;
        const result = await this.send(`interact/${this.versionID}`, {
            body: { state: this.state, request: action, config },
            method: 'POST',
            headers: { sessionID, platform: 'chat-prototype' },
        });
        this.state = result.state;
        return result;
    }
    async feedback() {
        return Promise.reject(new Error('not implemented'));
    }
    // TODO: expose authenticated publishing
    getPublishing() {
        return Promise.reject(new Error('not implemented'));
    }
    // TODO: expose authenticated createTranscript
    createTranscript() {
        return Promise.reject(new Error('not implemented'));
    }
}
