// @actbase/native/lib/devtools/hook/useConsole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tslib = require("tslib");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var prevData = {};

var parseLogs = function parseLogs(data, depth) {
  var _a;

  if (depth === void 0) {
    depth = 0;
  }

  if (!data) return data;

  if (depth === 0) {
    return Object.values(data).map(function (v) {
      if (typeof v === 'string' || typeof v === 'number') {
        return v;
      }

      return JSON.stringify(parseLogs(v, 1), null, 2);
    }).join(' ').replace(/"\$object\$"/g, '[Object object]').replace(/"\$Object\$"/g, '[Object object]').replace(/"\$function\$"/g, '[Function]').replace(/"\$Function\$"/g, '[Function]').replace(/"\$Array\$"/g, '[Array]');
  }

  if (typeof data === 'function') return '$Function$';
  if (_typeof(data) !== 'object') return data;
  if (depth >= 3 && Array.isArray(data)) return '$Array$';
  if (depth >= 3) return "$" + _typeof(data) + "$";

  if (Array.isArray(data)) {
    return data === null || data === void 0 ? void 0 : data.map(function (v) {
      return parseLogs(v, depth + 1);
    });
  }

  return (_a = Object.keys(data)) === null || _a === void 0 ? void 0 : _a.reduce(function (a, v) {
    a[v] = parseLogs(data[v], depth + 1);
    return a;
  }, {});
};

var useConsole = function useConsole(enabled, types, setConsoleLogs) {
  if (types === void 0) {
    types = ['log'];
  }

  _react["default"].useEffect(function () {
    if (!enabled) return undefined;
    types.forEach(function (type) {
      prevData[type] = console[type];

      console[type] = function console() {
        var _a, _b;

        var body = parseLogs(arguments, 0);
        setConsoleLogs(function (o) {
          var draft = (0, _tslib.__spreadArrays)(o);

          if (draft.length > 100) {
            draft.splice(100, draft.length - 100);
          }

          draft.splice(0, 0, {
            type: 'log',
            body: body,
            time: new Date().getTime()
          });
          return draft;
        });
        (_b = (_a = prevData[type]).apply) === null || _b === void 0 ? void 0 : _b.call(_a, this, arguments);
      };
    });
    return function () {
      types.forEach(function (type) {
        console[type] = prevData[type];
        prevData[type] = undefined;
      });
    };
  }, [enabled]);
};

var _default = useConsole;
exports["default"] = _default;
