// @actbase/native/lib/devtools/hook/useNetwork.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tslib = require("tslib");

var _react = _interopRequireDefault(require("react"));

var _common = require("../common");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/* eslint-disable */
var prevData = {
  httpOpen: undefined,
  httpSend: undefined
};

var useNetwork = function useNetwork(enabled, setHttpLogs) {
  _react["default"].useEffect(function () {
    if (!enabled) return undefined;
    prevData.httpSend = XMLHttpRequest.prototype.send;

    XMLHttpRequest.prototype.send = function send(body) {
      var _a, _b;

      if (body) {
        var that_1 = this;
        var _method_1 = that_1._method,
            _url_1 = that_1._url;
        if (_url_1.endsWith('/symbolicate') && _url_1.startsWith('http://localhost') && _method_1 === 'POST') return;
        setHttpLogs(function (o) {
          var draft = (0, _tslib.__spreadArrays)(o);

          if (draft.length > 100) {
            draft.splice(100, draft.length - 100);
          }

          var ix = draft.findIndex(function (x) {
            return x.obj === that_1;
          });

          if (ix < 0) {
            draft.splice(0, 0, {
              obj: that_1,
              url: _url_1,
              method: _method_1,
              state: 'ready',
              time: new Date().getTime(),
              body: body
            });
          } else {
            draft[ix].body = body;
          }

          return draft;
        });
      }

      (_b = (_a = prevData.httpSend) === null || _a === void 0 ? void 0 : _a.apply) === null || _b === void 0 ? void 0 : _b.call(_a, this, arguments);
    };

    prevData.httpOpen = XMLHttpRequest.prototype.open;

    XMLHttpRequest.prototype.open = function open() {
      var _a, _b;

      this.addEventListener('readystatechange', function readystatechange() {
        var that = this;

        var _a = this,
            readyState = _a.readyState,
            status = _a.status,
            _method = _a._method,
            _url = _a._url,
            _response = _a._response,
            _headers = _a._headers;

        var success = status === 0 || status >= 200 && status < 400;
        if (_url.endsWith('/symbolicate') && _url.startsWith('http://localhost') && _method === 'POST') return;
        setHttpLogs(function (o) {
          var draft = (0, _tslib.__spreadArrays)(o);
          if (draft.length > 100) draft.splice(100, o.length - 100);
          var ix = draft.findIndex(function (x) {
            return x.obj === that;
          });

          if (readyState === _common.READY_STATUS.OPENED) {
            if (ix < 0) {
              draft.splice(0, 0, {
                obj: that,
                url: _url,
                method: _method,
                state: 'ready',
                time: new Date().getTime()
              });
            } else {
              draft[ix].state = 'ready';
              draft[ix].time = new Date().getTime();
            }
          } else if (readyState === _common.READY_STATUS.DONE && ix >= 0) {
            draft[ix].state = success ? 'ok' : 'fail';
            draft[ix].status = status;
            draft[ix].finish = new Date().getTime();
            draft[ix].obj = undefined;
            draft[ix].response = _response;
            draft[ix].headers = _headers;
          }

          return draft;
        });
      }, false);
      (_b = (_a = prevData.httpOpen) === null || _a === void 0 ? void 0 : _a.apply) === null || _b === void 0 ? void 0 : _b.call(_a, this, arguments);
    };

    return function () {
      XMLHttpRequest.prototype.open = prevData.httpOpen;
      prevData.httpOpen = undefined;
      XMLHttpRequest.prototype.send = prevData.httpSend;
      prevData.httpSend = undefined;
    };
  }, [enabled]);
};

var _default = useNetwork;
exports["default"] = _default;
