// @posthog/wizard/dist/src/utils/logging.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareMessage = prepareMessage;
exports.l = l;
exports.nl = nl;
exports.green = green;
exports.red = red;
exports.dim = dim;
exports.yellow = yellow;
exports.cyan = cyan;
exports.debug = debug;
const chalk_1 = __importDefault(require("chalk"));
function prepareMessage(msg) {
    if (typeof msg === 'string') {
        return msg;
    }
    if (msg instanceof Error) {
        return `${msg.stack || ''}`;
    }
    return JSON.stringify(msg, null, '\t');
}
function l(msg) {
    // eslint-disable-next-line no-console
    console.log(msg);
}
function nl() {
    return l('');
}
function green(msg) {
    return l(chalk_1.default.green(prepareMessage(msg)));
}
function red(msg) {
    return l(chalk_1.default.red(prepareMessage(msg)));
}
function dim(msg) {
    return l(chalk_1.default.dim(prepareMessage(msg)));
}
function yellow(msg) {
    return l(chalk_1.default.yellow(prepareMessage(msg)));
}
function cyan(msg) {
    return l(chalk_1.default.cyan(prepareMessage(msg)));
}
function debug(msg) {
    return l(chalk_1.default.italic.yellow(prepareMessage(msg)));
}
