// @posthog/wizard/dist/src/steps/upload-environment-variables/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadEnvironmentVariablesStep = void 0;
const telemetry_1 = require("../../telemetry");
const analytics_1 = require("../../utils/analytics");
const clack_1 = __importDefault(require("../../utils/clack"));
const clack_utils_1 = require("../../utils/clack-utils");
const vercel_1 = require("./providers/vercel");
const uploadEnvironmentVariablesStep = async (envVars, { integration, options, }) => {
    const providers = [
        new vercel_1.VercelEnvironmentProvider(options),
    ];
    let provider = null;
    for (const p of providers) {
        if (await p.detect()) {
            provider = p;
            break;
        }
    }
    if (!provider) {
        analytics_1.analytics.capture('wizard interaction', {
            action: 'not uploading environment variables',
            reason: 'no environment provider found',
            integration,
        });
        return [];
    }
    const upload = await (0, clack_utils_1.abortIfCancelled)(clack_1.default.select({
        message: `It looks like you are using ${provider.name}. Would you like to upload the environment variables?`,
        options: [
            {
                value: true,
                label: 'Yes',
                hint: `Upload the environment variables to ${provider.name}`,
            },
            {
                value: false,
                label: 'No',
                hint: `Skip uploading environment variables to ${provider.name} - you can do this later`,
            },
        ],
    }), integration);
    if (!upload) {
        analytics_1.analytics.capture('wizard interaction', {
            action: 'not uploading environment variables',
            reason: 'user declined to upload',
            provider: provider.name,
            integration,
        });
        return [];
    }
    const results = await (0, telemetry_1.traceStep)('uploading environment variables', async () => {
        return await provider.uploadEnvVars(envVars);
    });
    analytics_1.analytics.capture('wizard interaction', {
        action: 'uploaded environment variables',
        provider: provider.name,
        integration,
    });
    return Object.keys(results).filter((key) => results[key]);
};
exports.uploadEnvironmentVariablesStep = uploadEnvironmentVariablesStep;
