// @voiceflow/base-types/build/cjs/version/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultSettings = exports.GlobalNoMatchType = exports.SessionType = exports.RepeatType = void 0;
var RepeatType;
(function (RepeatType) {
    RepeatType["OFF"] = "OFF";
    RepeatType["ALL"] = "ALL";
    RepeatType["DIALOG"] = "DIALOG";
})(RepeatType || (exports.RepeatType = RepeatType = {}));
var SessionType;
(function (SessionType) {
    SessionType["RESUME"] = "resume";
    SessionType["RESTART"] = "restart";
})(SessionType || (exports.SessionType = SessionType = {}));
var GlobalNoMatchType;
(function (GlobalNoMatchType) {
    GlobalNoMatchType["STATIC"] = "static";
    GlobalNoMatchType["GENERATIVE"] = "generative";
})(GlobalNoMatchType || (exports.GlobalNoMatchType = GlobalNoMatchType = {}));
const defaultSettings = ({ error = null, repeat = RepeatType.ALL, session = { type: SessionType.RESTART }, defaultVoice = null, defaultCanvasNodeVisibility = null, defaultCarouselLayout = null, globalNoMatch = { type: GlobalNoMatchType.STATIC, prompt: undefined }, globalNoReply = { delay: undefined, prompt: undefined }, intentConfidence = 0.6, } = {}) => ({
    error,
    repeat,
    session,
    defaultVoice,
    defaultCanvasNodeVisibility,
    defaultCarouselLayout,
    globalNoMatch,
    globalNoReply,
    intentConfidence,
});
exports.defaultSettings = defaultSettings;
