// @voiceflow/base-types/build/esm/models/project/knowledgeBase.js
export var KnowledgeBaseDocumentType;
(function (KnowledgeBaseDocumentType) {
    KnowledgeBaseDocumentType["PDF"] = "pdf";
    KnowledgeBaseDocumentType["TEXT"] = "text";
    KnowledgeBaseDocumentType["URL"] = "url";
    KnowledgeBaseDocumentType["DOCX"] = "docx";
    KnowledgeBaseDocumentType["TABLE"] = "table";
})(KnowledgeBaseDocumentType || (KnowledgeBaseDocumentType = {}));
export var KnowledgeBaseDocumentRefreshRate;
(function (KnowledgeBaseDocumentRefreshRate) {
    KnowledgeBaseDocumentRefreshRate["DAILY"] = "daily";
    KnowledgeBaseDocumentRefreshRate["WEEKLY"] = "weekly";
    KnowledgeBaseDocumentRefreshRate["MONTHLY"] = "monthly";
    KnowledgeBaseDocumentRefreshRate["NEVER"] = "never";
})(KnowledgeBaseDocumentRefreshRate || (KnowledgeBaseDocumentRefreshRate = {}));
export var KnowledgeBaseBooleanOperators;
(function (KnowledgeBaseBooleanOperators) {
    KnowledgeBaseBooleanOperators["AND"] = "and";
    KnowledgeBaseBooleanOperators["OR"] = "or";
})(KnowledgeBaseBooleanOperators || (KnowledgeBaseBooleanOperators = {}));
export var IntegrationTypes;
(function (IntegrationTypes) {
    IntegrationTypes["ZENDESK"] = "zendesk";
})(IntegrationTypes || (IntegrationTypes = {}));
export var KnowledgeBaseDocumentStatus;
(function (KnowledgeBaseDocumentStatus) {
    KnowledgeBaseDocumentStatus["ERROR"] = "ERROR";
    KnowledgeBaseDocumentStatus["SUCCESS"] = "SUCCESS";
    KnowledgeBaseDocumentStatus["PENDING"] = "PENDING";
    KnowledgeBaseDocumentStatus["INITIALIZED"] = "INITIALIZED";
})(KnowledgeBaseDocumentStatus || (KnowledgeBaseDocumentStatus = {}));
export var ChunkStrategyType;
(function (ChunkStrategyType) {
    ChunkStrategyType["RECURSIVE_TEXT_SPLITTER"] = "recursive_text_splitter";
})(ChunkStrategyType || (ChunkStrategyType = {}));
