// @actbase/native/lib/elems/Touchable.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _Actbase = _interopRequireDefault(require("../Actbase"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var defaults = _Actbase["default"].defaults;

var Touchable = function Touchable(props) {
  var renderComponent = props.renderComponent,
      onProcess = props.onProcess;
  var Element = renderComponent || _reactNative.TouchableOpacity;

  var _a = (0, _react.useState)(false),
      lock = _a[0],
      setLock = _a[1];

  var handlePress = function handlePress(e) {
    var _a;

    if (!(props === null || props === void 0 ? void 0 : props.onPress)) return;
    var pressed = (_a = props === null || props === void 0 ? void 0 : props.onPress) === null || _a === void 0 ? void 0 : _a.call(props, e);

    if (pressed instanceof Promise) {
      setLock(true);
      if (onProcess) onProcess(true);
      pressed["finally"](function () {
        setLock(false);
        if (onProcess) onProcess(false);
      });
    } else {
      setLock(true);
      setTimeout(function () {
        return setLock(false);
      }, defaults.Touchable.waitDelay);
    }
  };

  return _react["default"].createElement(Element, (0, _tslib.__assign)({
    activeOpacity: defaults.Touchable.activeOpacity
  }, props, {
    onPress: handlePress,
    disabled: lock || (props === null || props === void 0 ? void 0 : props.disabled)
  }));
};

Touchable.defaultProps = {
  onPress: function onPress() {
    return undefined;
  },
  onProcess: function onProcess() {
    return undefined;
  }
};
var _default = Touchable;
exports["default"] = _default;
