// @voiceflow/base-types/build/cjs/models/project/knowledgeBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChunkStrategyType = exports.KnowledgeBaseDocumentStatus = exports.IntegrationTypes = exports.KnowledgeBaseBooleanOperators = exports.KnowledgeBaseDocumentRefreshRate = exports.KnowledgeBaseDocumentType = void 0;
var KnowledgeBaseDocumentType;
(function (KnowledgeBaseDocumentType) {
    KnowledgeBaseDocumentType["PDF"] = "pdf";
    KnowledgeBaseDocumentType["TEXT"] = "text";
    KnowledgeBaseDocumentType["URL"] = "url";
    KnowledgeBaseDocumentType["DOCX"] = "docx";
    KnowledgeBaseDocumentType["TABLE"] = "table";
})(KnowledgeBaseDocumentType || (exports.KnowledgeBaseDocumentType = KnowledgeBaseDocumentType = {}));
var KnowledgeBaseDocumentRefreshRate;
(function (KnowledgeBaseDocumentRefreshRate) {
    KnowledgeBaseDocumentRefreshRate["DAILY"] = "daily";
    KnowledgeBaseDocumentRefreshRate["WEEKLY"] = "weekly";
    KnowledgeBaseDocumentRefreshRate["MONTHLY"] = "monthly";
    KnowledgeBaseDocumentRefreshRate["NEVER"] = "never";
})(KnowledgeBaseDocumentRefreshRate || (exports.KnowledgeBaseDocumentRefreshRate = KnowledgeBaseDocumentRefreshRate = {}));
var KnowledgeBaseBooleanOperators;
(function (KnowledgeBaseBooleanOperators) {
    KnowledgeBaseBooleanOperators["AND"] = "and";
    KnowledgeBaseBooleanOperators["OR"] = "or";
})(KnowledgeBaseBooleanOperators || (exports.KnowledgeBaseBooleanOperators = KnowledgeBaseBooleanOperators = {}));
var IntegrationTypes;
(function (IntegrationTypes) {
    IntegrationTypes["ZENDESK"] = "zendesk";
})(IntegrationTypes || (exports.IntegrationTypes = IntegrationTypes = {}));
var KnowledgeBaseDocumentStatus;
(function (KnowledgeBaseDocumentStatus) {
    KnowledgeBaseDocumentStatus["ERROR"] = "ERROR";
    KnowledgeBaseDocumentStatus["SUCCESS"] = "SUCCESS";
    KnowledgeBaseDocumentStatus["PENDING"] = "PENDING";
    KnowledgeBaseDocumentStatus["INITIALIZED"] = "INITIALIZED";
})(KnowledgeBaseDocumentStatus || (exports.KnowledgeBaseDocumentStatus = KnowledgeBaseDocumentStatus = {}));
var ChunkStrategyType;
(function (ChunkStrategyType) {
    ChunkStrategyType["RECURSIVE_TEXT_SPLITTER"] = "recursive_text_splitter";
})(ChunkStrategyType || (exports.ChunkStrategyType = ChunkStrategyType = {}));
