// @actbase/native/lib/forms/inputs/Text.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _Context = require("../Context");

var _ref = require("../../utils/ref");

var _styles = require("../../utils/styles");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var InputText = _react["default"].forwardRef(function (props, fowarededRef) {
  var _a, _b, _c;

  var ref = (0, _react.useRef)(null);
  var remote = (0, _Context.useFormField)((0, _tslib.__assign)((0, _tslib.__assign)({}, props), {
    ref: ref
  }));

  var _d = (0, _react.useState)(props.disabled || props.readonly ? 'disabled' : 'idle'),
      state = _d[0],
      setState = _d[1];

  var before = props.before,
      after = props.after,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      rules = props.rules,
      editable = props.editable,
      disabled = props.disabled,
      style = props.style,
      inputProps = (0, _tslib.__rest)(props, ["before", "after", "onFocus", "onBlur", "rules", "editable", "disabled", "style"]);
  (0, _react.useEffect)(function () {
    var _a;

    (_a = remote.setValue) === null || _a === void 0 ? void 0 : _a.call(remote, inputProps === null || inputProps === void 0 ? void 0 : inputProps.value);
  }, [inputProps === null || inputProps === void 0 ? void 0 : inputProps.value]);

  var handleFocus = function handleFocus(e) {
    setState('focus');
    onFocus === null || onFocus === void 0 ? void 0 : onFocus(e);
  };

  var handleBlur = function handleBlur(e) {
    setState('idle');
    onBlur === null || onBlur === void 0 ? void 0 : onBlur(e);
  };

  var handleChangeText = function handleChangeText(text) {
    return (0, _tslib.__awaiter)(void 0, void 0, void 0, function () {
      var value, frmValues, anyRules, promises;

      var _a, _b;

      return (0, _tslib.__generator)(this, function (_c) {
        (_a = inputProps === null || inputProps === void 0 ? void 0 : inputProps.onChangeText) === null || _a === void 0 ? void 0 : _a.call(inputProps, text);
        value = text;
        frmValues = (_b = remote.setValue) === null || _b === void 0 ? void 0 : _b.call(remote, value);

        if (rules) {
          anyRules = rules;

          if (!Array.isArray(anyRules)) {
            anyRules = [anyRules];
          }

          promises = rules === null || rules === void 0 ? void 0 : rules.map(function (v) {
            return v(value, frmValues);
          });
          Promise.all(promises).then(function () {
            var _a;

            return (_a = props.onError) === null || _a === void 0 ? void 0 : _a.call(props, undefined);
          })["catch"](props.onError);
        }

        return [2
        /*return*/
        ];
      });
    });
  };

  console.log(remote.props, state);
  var inputStyle = (0, _react.useMemo)(function () {
    var args = {
      // padding: 0,
      // paddingTop: 0,
      // paddingLeft: 0,
      // paddingStart: 0,
      // paddingRight: 0,
      // paddingBottom: 0,
      // paddingEnd: 0,
      height: '100%',
      flex: 1
    };
    Object.keys(style || {}).forEach(function (key) {
      if (style && _styles.TEXT_STYLE_NAMES.includes(key)) {
        // @ts-ignore
        args[key] = style[key];
      }
    });
    return args;
  }, [style]);
  return _react["default"].createElement(_reactNative.View, {
    style: [style]
  }, before, _react["default"].createElement(_reactNative.TextInput, (0, _tslib.__assign)({
    ref: (0, _ref.composeRef)(fowarededRef, ref)
  }, inputProps, {
    style: inputStyle,
    onChangeText: handleChangeText,
    onBlur: handleBlur,
    onFocus: handleFocus,
    editable: !(!editable || disabled || ((_a = remote.props) === null || _a === void 0 ? void 0 : _a.submitting)),
    returnKeyType: inputProps.multiline ? undefined : inputProps.returnKeyType || ((_b = remote.props) === null || _b === void 0 ? void 0 : _b.returnKeyType),
    onSubmitEditing: inputProps.multiline ? undefined : inputProps.onSubmitEditing || ((_c = remote.props) === null || _c === void 0 ? void 0 : _c.onSubmitEditing)
  })), after);
});

InputText.defaultProps = {
  editable: true,
  autoCorrect: false,
  clearButtonMode: 'always'
};
var _default = InputText;
exports["default"] = _default;
