// @actbase/native/lib/Application.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Application = exports.Providers = void 0;

var _tslib = require("tslib");

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _reactNativeSafeAreaContext = require("react-native-safe-area-context");

var _reactAbsolute = require("@actbase/react-absolute");

var _devtools = _interopRequireDefault(require("./devtools"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var $Providers = function $Providers() {
  return _react["default"].createElement(_react["default"].Fragment, null);
};

$Providers.$$NAME = '__@@ABX_PROVID@@';

var AppElement = function AppElement(_a) {
  var _b;

  var _c, _d, _e;

  var children = _a.children,
      style = _a.style,
      debug = _a.debug,
      reduxStore = _a.reduxStore;

  var oChildren = _react["default"].Children.toArray(children);

  var providerGroup = oChildren === null || oChildren === void 0 ? void 0 : oChildren.find(function (el) {
    var _a;

    return ((_a = el.type) === null || _a === void 0 ? void 0 : _a.$$NAME) === ($Providers === null || $Providers === void 0 ? void 0 : $Providers.$$NAME);
  });
  var bodies = oChildren === null || oChildren === void 0 ? void 0 : oChildren.filter(function (el) {
    var _a;

    return ((_a = el.type) === null || _a === void 0 ? void 0 : _a.$$NAME) !== ($Providers === null || $Providers === void 0 ? void 0 : $Providers.$$NAME);
  });
  var providers = providerGroup ? _react["default"].Children.toArray(providerGroup.props.children) : [];
  var exports = (_c = providers === null || providers === void 0 ? void 0 : providers.reduce) === null || _c === void 0 ? void 0 : _c.call(providers, function (el, provider, index) {
    var _a, _b;

    var last = el;

    while (((_a = last.props.children) === null || _a === void 0 ? void 0 : _a.length) > 0) {
      last = (_b = last.props.children) === null || _b === void 0 ? void 0 : _b[0];
    }

    last.props.children.push(_react["default"].createElement(provider.type, (0, _tslib.__assign)({
      key: "" + index
    }, provider.props), []));
    return el;
  }, _react["default"].createElement(_react["default"].Fragment, null, []));
  var output;

  if (exports) {
    var last = exports;

    while (((_d = last.props.children) === null || _d === void 0 ? void 0 : _d.length) > 0) {
      last = (_e = last.props.children) === null || _e === void 0 ? void 0 : _e[0];
    }

    (_b = last.props.children).push.apply(_b, bodies);

    output = exports;
  } else {
    output = bodies;
  }

  return _react["default"].createElement(_reactNativeSafeAreaContext.SafeAreaProvider, null, _react["default"].createElement(_devtools["default"], {
    debug: debug,
    reduxStore: reduxStore
  }, _react["default"].createElement(_reactNative.View, {
    style: style
  }, _react["default"].createElement(_reactAbsolute.AbsoluteProvider, null, output))));
};

AppElement.defaultProps = {
  debug: __DEV__,
  style: {
    flex: 1
  }
};
var Providers = $Providers;
exports.Providers = Providers;

var Application = _react["default"].memo(AppElement);

exports.Application = Application;
