// @actbase/native/lib/utils/size.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.measure = exports.getWindowSize = void 0;

var _reactNative = require("react-native");

var getWindowSize = function getWindowSize() {
  return _reactNative.Dimensions.get('screen');
};

exports.getWindowSize = getWindowSize;

var measure = function measure(target) {
  return new Promise(function (resolve, reject) {
    try {
      var node = (0, _reactNative.findNodeHandle)(target);
      if (node === null) throw new Error();

      _reactNative.UIManager.measure(node, function (originX, originY, width, oHeight, pageX, pageY) {
        return resolve({
          originX: originX,
          originY: originY,
          width: width,
          height: oHeight,
          pageX: pageX,
          pageY: pageY
        });
      });
    } catch (e) {
      reject(e);
    }
  });
};

exports.measure = measure;
var _default = {
  getWindowSize: getWindowSize,
  measure: measure
};
exports["default"] = _default;
