// @posthog/wizard/dist/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WIZARD_INTERACTION_EVENT_NAME = exports.OAUTH_PORT = exports.POSTHOG_DEV_CLIENT_ID = exports.POSTHOG_EU_CLIENT_ID = exports.POSTHOG_US_CLIENT_ID = exports.DUMMY_PROJECT_API_KEY = exports.ANALYTICS_HOST_URL = exports.ANALYTICS_POSTHOG_PUBLIC_PROJECT_WRITE_KEY = exports.DEFAULT_HOST_URL = exports.ISSUES_URL = exports.DEFAULT_URL = exports.DEBUG = exports.IS_DEV = exports.WizardVariant = exports.FeatureFlagDefinition = exports.Integration = void 0;
exports.getIntegrationDescription = getIntegrationDescription;
exports.getIntegrationChoices = getIntegrationChoices;
var Integration;
(function (Integration) {
    Integration["nextjs"] = "nextjs";
    Integration["react"] = "react";
    Integration["svelte"] = "svelte";
    Integration["reactNative"] = "react-native";
    Integration["astro"] = "astro";
})(Integration || (exports.Integration = Integration = {}));
var FeatureFlagDefinition;
(function (FeatureFlagDefinition) {
    FeatureFlagDefinition["NextV2"] = "wizard-next-v2";
})(FeatureFlagDefinition || (exports.FeatureFlagDefinition = FeatureFlagDefinition = {}));
var WizardVariant;
(function (WizardVariant) {
    WizardVariant["Legacy"] = "legacy";
    WizardVariant["Agent"] = "agent";
})(WizardVariant || (exports.WizardVariant = WizardVariant = {}));
function getIntegrationDescription(type) {
    switch (type) {
        case Integration.nextjs:
            return 'Next.js';
        case Integration.react:
            return 'React';
        case Integration.reactNative:
            return 'React Native';
        case Integration.svelte:
            return 'Svelte';
        case Integration.astro:
            return 'Astro';
        default:
            throw new Error(`Unknown integration ${type}`);
    }
}
function getIntegrationChoices() {
    return Object.keys(Integration).map((type) => ({
        name: getIntegrationDescription(type),
        value: type,
    }));
}
exports.IS_DEV = ['test', 'development'].includes(process.env.NODE_ENV ?? '');
exports.DEBUG = false;
exports.DEFAULT_URL = exports.IS_DEV
    ? 'http://localhost:8010'
    : 'https://us.posthog.com';
exports.ISSUES_URL = 'https://github.com/posthog/wizard/issues';
exports.DEFAULT_HOST_URL = exports.IS_DEV
    ? 'http://localhost:8010'
    : 'https://us.i.posthog.com';
exports.ANALYTICS_POSTHOG_PUBLIC_PROJECT_WRITE_KEY = 'sTMFPsFhdP1Ssg';
exports.ANALYTICS_HOST_URL = 'https://internal-j.posthog.com';
exports.DUMMY_PROJECT_API_KEY = '_YOUR_POSTHOG_PROJECT_API_KEY_';
exports.POSTHOG_US_CLIENT_ID = 'c4Rdw8DIxgtQfA80IiSnGKlNX8QN00cFWF00QQhM';
exports.POSTHOG_EU_CLIENT_ID = 'bx2C5sZRN03TkdjraCcetvQFPGH6N2Y9vRLkcKEy';
exports.POSTHOG_DEV_CLIENT_ID = 'DC5uRLVbGI02YQ82grxgnK6Qn12SXWpCqdPb60oZ';
exports.OAUTH_PORT = 8239;
exports.WIZARD_INTERACTION_EVENT_NAME = 'wizard interaction';
