// @voiceflow/voiceflow-types/build/cjs/constants/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isVoiceflowLanguage = exports.Language = exports.Locale = exports.Voice = void 0;
var Voice;
(function (Voice) {
    Voice["AUDIO"] = "audio";
    Voice["DEFAULT"] = "default";
})(Voice || (exports.Voice = Voice = {}));
var Locale;
(function (Locale) {
    // English
    Locale["EN_US"] = "en-US";
    // Arabic
    Locale["AR_AR"] = "ar-AR";
    // Bulgarian
    Locale["BG_BG"] = "bg-BG";
    // Catalan
    Locale["CA_ES"] = "ca-ES";
    // Czech
    Locale["CS_CZ"] = "cs-CZ";
    // Chinese
    Locale["ZH_CN"] = "zh-CN";
    Locale["ZH_TW"] = "zh-TW";
    // Danish
    Locale["DA_DK"] = "da-DK";
    // Dutch / Flemish
    Locale["NL_NL"] = "nl-NL";
    Locale["NL_BE"] = "nl-BE";
    // Estonian
    Locale["ET_EE"] = "et-EE";
    // French
    Locale["FR_FR"] = "fr-FR";
    Locale["FR_CA"] = "fr-CA";
    // German
    Locale["DE_DE"] = "de-DE";
    // Hebrew
    Locale["HE_IL"] = "he-IL";
    // Hungarian
    Locale["HU_HU"] = "hu-HU";
    // Gujarati
    Locale["GU_IN"] = "gu-IN";
    // Hindi
    Locale["HI_IN"] = "hi-IN";
    // Italian
    Locale["IT_IT"] = "it-IT";
    // Japanese
    Locale["JA_JP"] = "ja-JP";
    // Korean
    Locale["KO_KR"] = "ko-KR";
    // Marathi
    Locale["MR_IN"] = "mr-IN";
    // Polish
    Locale["PL_PL"] = "pl-PL";
    // Portugese
    Locale["PT_PT"] = "pt-PT";
    Locale["PT_BR"] = "pt-BR";
    // Romanian
    Locale["RO_RO"] = "ro-RO";
    // Russian
    Locale["RU_RU"] = "ru-RU";
    // Spanish
    Locale["ES_ES"] = "es-ES";
    Locale["ES_MX"] = "es-MX";
    // Tamil
    Locale["TA_IN"] = "ta-IN";
    // Telugu
    Locale["TE_IN"] = "te-IN";
    // Turkish
    Locale["TR_TR"] = "tr-TR";
    // Ukrainian
    Locale["UK_UA"] = "uk-UA";
    // Vietnamese
    Locale["VI_VN"] = "vi-VN";
})(Locale || (exports.Locale = Locale = {}));
var Language;
(function (Language) {
    // English
    Language["EN"] = "en";
    // Arabic
    Language["AR"] = "ar";
    // Bulgarian
    Language["BG"] = "bg";
    // Catalan
    Language["CA"] = "ca";
    // Czech
    Language["CS"] = "cs";
    // Chinese
    Language["ZH"] = "zh";
    // Danish
    Language["DA"] = "da";
    // Dutch / Flemish
    Language["NL"] = "nl";
    // Estonian
    Language["ET"] = "et";
    // French
    Language["FR"] = "fr";
    // German
    Language["DE"] = "de";
    // Gujarati
    Language["GU"] = "gu";
    // Hebrew
    Language["HE"] = "he";
    // Hindi
    Language["HI"] = "hi";
    // Hungarian
    Language["HU"] = "hu";
    // Italian
    Language["IT"] = "it";
    // Japanese
    Language["JA"] = "ja";
    // Korean
    Language["KO"] = "ko";
    // Marathi
    Language["MR"] = "mr";
    // Polish
    Language["PL"] = "pl";
    // Portugese
    Language["PT"] = "pt";
    // Romanian
    Language["RO"] = "ro";
    // Russian
    Language["RU"] = "ru";
    // Spanish
    Language["ES"] = "es";
    // Tamil
    Language["TA"] = "ta";
    // Telugu
    Language["TE"] = "te";
    // Turkish
    Language["TR"] = "tr";
    // Ukrainian
    Language["UK"] = "uk";
    // Vietnamese
    Language["VI"] = "vi";
})(Language || (exports.Language = Language = {}));
const LanguageValues = new Set(Object.values(Language));
const isVoiceflowLanguage = (lang) => LanguageValues.has(lang);
exports.isVoiceflowLanguage = isVoiceflowLanguage;
