// @posthog/wizard/dist/src/utils/__tests__/clack-utils.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const clack_utils_1 = require("../clack-utils");
const ChildProcess = __importStar(require("node:child_process"));
jest.mock('node:child_process', () => ({
    __esModule: true,
    ...jest.requireActual('node:child_process'),
}));
jest.mock('@clack/prompts', () => ({
    log: {
        info: jest.fn(),
        success: jest.fn(),
        warn: jest.fn(),
        error: jest.fn(),
    },
    text: jest.fn(),
    confirm: jest.fn(),
    cancel: jest.fn(),
    outro: jest.fn(),
    // passthrough for abortIfCancelled
    isCancel: jest.fn().mockReturnValue(false),
    spinner: jest
        .fn()
        .mockImplementation(() => ({ start: jest.fn(), stop: jest.fn() })),
}));
describe.skip('installPackage', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    it('force-installs a package if the forceInstall flag is set', async () => {
        const packageManagerMock = {
            name: 'npm',
            label: 'NPM',
            installCommand: 'npm install',
            buildCommand: 'npm run build',
            runScriptCommand: 'npm run',
            flags: '',
            forceInstallFlag: '--force',
            detect: jest.fn(),
            addOverride: jest.fn(),
        };
        const execSpy = jest
            .spyOn(ChildProcess, 'exec')
            // @ts-expect-error - don't care about the return value
            .mockImplementationOnce((cmd, cb) => {
            if (cb) {
                // @ts-expect-error - don't care about the options value
                cb(null, '', '');
            }
        });
        await (0, clack_utils_1.installPackage)({
            alreadyInstalled: false,
            packageName: 'posthog-js',
            packageNameDisplayLabel: 'posthog-js',
            forceInstall: true,
            askBeforeUpdating: false,
            packageManager: packageManagerMock,
            installDir: process.cwd(),
        });
        expect(execSpy).toHaveBeenCalledWith('npm install posthog-js  --force', expect.any(Function));
    });
    it.each([false, undefined])("doesn't force-install a package if the forceInstall flag is %s", async (flag) => {
        const packageManagerMock = {
            name: 'npm',
            label: 'NPM',
            installCommand: 'npm install',
            buildCommand: 'npm run build',
            runScriptCommand: 'npm run',
            flags: '',
            forceInstallFlag: '--force',
            detect: jest.fn(),
            addOverride: jest.fn(),
        };
        const execSpy = jest
            .spyOn(ChildProcess, 'exec')
            // @ts-expect-error - don't care about the return value
            .mockImplementationOnce((cmd, cb) => {
            if (cb) {
                // @ts-expect-error - don't care about the options value
                cb(null, '', '');
            }
        });
        await (0, clack_utils_1.installPackage)({
            alreadyInstalled: false,
            packageName: 'posthog-js',
            packageNameDisplayLabel: 'posthog-js',
            forceInstall: flag,
            askBeforeUpdating: false,
            packageManager: packageManagerMock,
            installDir: process.cwd(),
        });
        expect(execSpy).toHaveBeenCalledWith('npm install posthog-js  ', expect.any(Function));
    });
});
