// @actbase/native/lib/forms/Context.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFormField = exports.FormContext = void 0;

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var FormContext = _react["default"].createContext({});

exports.FormContext = FormContext;

var useFormField = function useFormField(options) {
  if (options === void 0) {
    options = {};
  }

  var ref = (0, _react.useRef)({});
  var context = (0, _react.useContext)(FormContext);

  var _a = (0, _react.useState)({}),
      receiveProps = _a[0],
      setReceiveProps = _a[1];

  var onReceiveProps = function onReceiveProps(data) {
    setReceiveProps(data);
  };

  (0, _react.useEffect)(function () {
    var _a;

    var o = (_a = context === null || context === void 0 ? void 0 : context.subscribe) === null || _a === void 0 ? void 0 : _a.call(context, (0, _tslib.__assign)((0, _tslib.__assign)({}, options), {
      onReceiveProps: onReceiveProps,
      focus: function focus() {
        var _a;

        return (options === null || options === void 0 ? void 0 : options.focus()) || ((_a = options === null || options === void 0 ? void 0 : options.ref) === null || _a === void 0 ? void 0 : _a.focus());
      },
      blur: function blur() {
        var _a;

        return (options === null || options === void 0 ? void 0 : options.blur()) || ((_a = options === null || options === void 0 ? void 0 : options.ref) === null || _a === void 0 ? void 0 : _a.blur());
      }
    }));
    ref.current.setValue = o.setValue;
    return o.unsubscribe;
  }, []);
  return (0, _tslib.__assign)((0, _tslib.__assign)({}, ref.current), {
    props: receiveProps
  });
};

exports.useFormField = useFormField;
