// @actbase/native/lib/forms/Form.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tslib = require("tslib");

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _size = require("../utils/size");

var _Context = require("./Context");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Form = function Form(props) {
  var onLayout = props.onLayout,
      oProps = (0, _tslib.__rest)(props, ["onLayout"]);
  var items = React.useRef({});
  var inValue = React.useRef({});

  var _a = React.useState(),
      lastLayout = _a[0],
      setLastLayout = _a[1];

  var _b = (0, React.useState)(0),
      lastIndex = _b[0],
      setLastIndex = _b[1];

  var getValues = function getValues() {
    var _a;

    var output = {};
    var keys = (_a = Object.keys(items === null || items === void 0 ? void 0 : items.current)) === null || _a === void 0 ? void 0 : _a.filter(function (v) {
      var _a;

      return !!((_a = items === null || items === void 0 ? void 0 : items.current[v]) === null || _a === void 0 ? void 0 : _a.name);
    });
    keys.forEach(function (key) {
      var option = items === null || items === void 0 ? void 0 : items.current[key];
      var name = (option === null || option === void 0 ? void 0 : option.name) || '';
      var multi = (option === null || option === void 0 ? void 0 : option.multiable) || (option === null || option === void 0 ? void 0 : option.multiable) === undefined && (option === null || option === void 0 ? void 0 : option.name) && Object.values(items === null || items === void 0 ? void 0 : items.current).filter(function (v) {
        return (v === null || v === void 0 ? void 0 : v.name) === (option === null || option === void 0 ? void 0 : option.name);
      }).length > 1;

      if (multi) {
        if (!output[name]) output[name] = [];
        output[name].push(inValue.current[key]);
      } else {
        output[name] = inValue.current[key];
      }
    });
    return output;
  };

  var submit = function submit() {
    getValues();
  };

  var remapFields = function remapFields() {
    return (0, _tslib.__awaiter)(void 0, void 0, void 0, function () {
      var elements;
      return (0, _tslib.__generator)(this, function (_a) {
        switch (_a.label) {
          case 0:
            return [4
            /*yield*/
            , Promise.all(Object.keys(items.current).map(function (key) {
              return (0, _tslib.__awaiter)(void 0, void 0, void 0, function () {
                var el, pos;

                var _a, _b;

                return (0, _tslib.__generator)(this, function (_c) {
                  switch (_c.label) {
                    case 0:
                      el = items === null || items === void 0 ? void 0 : items.current[key];
                      if (!(el && ((_a = el === null || el === void 0 ? void 0 : el.ref) === null || _a === void 0 ? void 0 : _a.current))) return [3
                      /*break*/
                      , 2];
                      return [4
                      /*yield*/
                      , (0, _size.measure)((_b = el.ref) === null || _b === void 0 ? void 0 : _b.current)];

                    case 1:
                      pos = _c.sent();

                      if ('pageY' in pos && 'pageX' in pos) {
                        el.area = -parseFloat(Math.floor(pos === null || pos === void 0 ? void 0 : pos.pageY) + "." + Math.floor(pos === null || pos === void 0 ? void 0 : pos.pageX));
                      }

                      _c.label = 2;

                    case 2:
                      return [2
                      /*return*/
                      ];
                  }
                });
              });
            }))];

          case 1:
            _a.sent();

            elements = Object.values(items.current).filter(function (v) {
              return !!v;
            }).sort(function (a, b) {
              if (((a === null || a === void 0 ? void 0 : a.area) || 0) < ((b === null || b === void 0 ? void 0 : b.area) || 0)) return 1;
              if (((a === null || a === void 0 ? void 0 : a.area) || 0) > ((b === null || b === void 0 ? void 0 : b.area) || 0)) return -1;
              return 0;
            });

            if (elements === null || elements === void 0 ? void 0 : elements.forEach) {
              elements.forEach(function (v, index) {
                var _a;

                if (!v) return;
                var args = {};

                if (elements.length - 1 <= index) {
                  args.returnKeyType = 'done';
                  args.onSubmitEditing = submit;
                } else {
                  args.returnKeyType = 'next';
                  args.onSubmitEditing = (_a = elements[index + 1]) === null || _a === void 0 ? void 0 : _a.focus;
                }

                if (v.onReceiveProps) v.onReceiveProps(args);
              });
            }

            return [2
            /*return*/
            ];
        }
      });
    });
  };

  var handleLayout = function handleLayout(e) {
    return (0, _tslib.__awaiter)(void 0, void 0, void 0, function () {
      var _a, width, height, pos;

      return (0, _tslib.__generator)(this, function (_b) {
        switch (_b.label) {
          case 0:
            if (onLayout) {
              onLayout(e);
            }

            _a = e.nativeEvent.layout, width = _a.width, height = _a.height;
            pos = {
              width: width,
              height: height
            };
            if ((lastLayout === null || lastLayout === void 0 ? void 0 : lastLayout.width) === pos.width && (lastLayout === null || lastLayout === void 0 ? void 0 : lastLayout.height) === pos.height) return [2
            /*return*/
            ];
            setLastLayout(pos);
            return [4
            /*yield*/
            , remapFields()];

          case 1:
            _b.sent();

            return [2
            /*return*/
            ];
        }
      });
    });
  };

  (0, React.useEffect)(function () {
    remapFields()["catch"](function () {
      return null;
    });
  }, [lastIndex]);

  var _unsubscribe = (0, React.useCallback)(function (idx) {
    items.current[idx] = undefined;
    inValue.current[idx] = undefined;
  }, []);

  var subscribe = (0, React.useCallback)(function (ssArgs) {
    setLastIndex(Object.keys(items.current || {}).length);
    var ix = (ssArgs === null || ssArgs === void 0 ? void 0 : ssArgs.idx) || String(Object.keys(items.current || {}).length + 1);
    items.current[ix] = ssArgs;
    return {
      unsubscribe: function unsubscribe() {
        return _unsubscribe(ix);
      },
      setValue: function setValue(args) {
        inValue.current[ix] = args;
        return getValues();
      }
    };
  }, []);
  var value = {
    subscribe: subscribe,
    submit: submit
  };
  return React.createElement(_Context.FormContext.Provider, {
    value: value
  }, React.createElement(_reactNative.View, (0, _tslib.__assign)({
    onLayout: handleLayout
  }, oProps)));
};

var _default = Form;
exports["default"] = _default;
