// @posthog/wizard/dist/src/nextjs/nextjs-wizard-agent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runNextjsWizardAgent = runNextjsWizardAgent;
/* Simplified Next.js wizard using posthog-agent with PostHog MCP */
const clack_utils_1 = require("../utils/clack-utils");
const package_json_1 = require("../utils/package-json");
const utils_1 = require("./utils");
const clack_1 = __importDefault(require("../utils/clack"));
const constants_1 = require("../lib/constants");
const analytics_1 = require("../utils/analytics");
const clack_utils_2 = require("../utils/clack-utils");
const urls_1 = require("../utils/urls");
const chalk_1 = __importDefault(require("chalk"));
const steps_1 = require("../steps");
const debug_1 = require("../utils/debug");
const agent_interface_1 = require("../lib/agent-interface");
/**
 * Simplified Next.js wizard that delegates to PostHog MCP's /integrate command
 */
async function runNextjsWizardAgent(options) {
    if (options.debug) {
        (0, debug_1.enableDebugLogs)();
    }
    (0, clack_utils_1.printWelcome)({
        wizardName: 'PostHog Next.js wizard (agent-powered)',
    });
    clack_1.default.log.info('🧙 The wizard has chosen you to try the next-generation agent integration for Next.js.\n\nStand by for the good stuff, and let me know how it goes:\n\ndanilo@posthog.com');
    const aiConsent = await (0, clack_utils_1.askForAIConsent)(options);
    if (!aiConsent) {
        await (0, clack_utils_1.abort)('This wizard uses an LLM agent to intelligently modify your project. Please view the docs to setup Next.js manually instead: https://posthog.com/docs/libraries/next-js', 0);
    }
    const cloudRegion = options.cloudRegion ?? (await (0, clack_utils_2.askForCloudRegion)());
    const typeScriptDetected = (0, clack_utils_1.isUsingTypeScript)(options);
    await (0, clack_utils_1.confirmContinueIfNoOrDirtyGitRepo)(options);
    const packageJson = await (0, clack_utils_1.getPackageDotJson)(options);
    await (0, clack_utils_1.ensurePackageIsInstalled)(packageJson, 'next', 'Next.js');
    const nextVersion = (0, package_json_1.getPackageVersion)('next', packageJson);
    analytics_1.analytics.setTag('nextjs-version', (0, utils_1.getNextJsVersionBucket)(nextVersion));
    analytics_1.analytics.capture(constants_1.WIZARD_INTERACTION_EVENT_NAME, {
        action: 'started agent integration',
    });
    const { projectApiKey, host, accessToken } = await (0, clack_utils_1.getOrAskForProjectData)({
        ...options,
        cloudRegion,
    });
    const router = await (0, utils_1.getNextJsRouter)(options);
    const routerType = router === utils_1.NextJsRouter.APP_ROUTER ? 'app' : 'pages';
    const spinner = clack_1.default.spinner();
    const agent = (0, agent_interface_1.initializeAgent)({
        workingDirectory: options.installDir,
        posthogMcpUrl: 'https://mcp.posthog.com/mcp',
        posthogApiKey: accessToken,
        debug: false,
    }, options, spinner);
    const integrationPrompt = buildIntegrationPrompt({
        framework: 'Next.js',
        version: nextVersion || 'latest',
        router: routerType,
        typescript: typeScriptDetected,
        projectApiKey,
        host,
    });
    await (0, agent_interface_1.runAgent)(agent, integrationPrompt, options, spinner, {
        estimatedDurationMinutes: 8,
        spinnerMessage: 'Writing your PostHog setup with events, error capture and more...',
        successMessage: 'PostHog integration complete',
        errorMessage: 'Integration failed',
    });
    const { relativeEnvFilePath, addedEnvVariables } = await (0, steps_1.addOrUpdateEnvironmentVariablesStep)({
        variables: {
            NEXT_PUBLIC_POSTHOG_KEY: projectApiKey,
            NEXT_PUBLIC_POSTHOG_HOST: host,
        },
        installDir: options.installDir,
        integration: constants_1.Integration.nextjs,
    });
    const uploadedEnvVars = await (0, steps_1.uploadEnvironmentVariablesStep)({
        NEXT_PUBLIC_POSTHOG_KEY: projectApiKey,
        NEXT_PUBLIC_POSTHOG_HOST: host,
    }, {
        integration: constants_1.Integration.nextjs,
        options,
    });
    await (0, steps_1.addMCPServerToClientsStep)({
        cloudRegion,
        integration: constants_1.Integration.nextjs,
    });
    const continueUrl = options.signup
        ? `${(0, urls_1.getCloudUrlFromRegion)(cloudRegion)}/products?source=wizard`
        : undefined;
    const changes = [
        addedEnvVariables
            ? `Added your Project API key to your ${relativeEnvFilePath} file`
            : '',
        uploadedEnvVars.length > 0
            ? `Uploaded your Project API key to your hosting provider`
            : '',
    ].filter(Boolean);
    const nextSteps = [
        uploadedEnvVars.length === 0
            ? `Upload your Project API key to your hosting provider`
            : '',
    ].filter(Boolean);
    const outroMessage = `
${chalk_1.default.green('Successfully installed PostHog!')}

${chalk_1.default.cyan('What the agent did:')}
• Analyzed your Next.js project structure (${routerType} router)
• Created and configured PostHog initializers
• Integrated PostHog into your application
${changes.map((change) => `• ${change}`).join('\n')}

${chalk_1.default.yellow('Next steps:')}
• Start your development server to see PostHog in action
• Visit your PostHog dashboard to see incoming events
${nextSteps.map((step) => `• ${step}`).join('\n')}

Learn more about PostHog + Next.js: ${chalk_1.default.cyan('https://posthog.com/docs/libraries/next-js')}
${continueUrl ? `\nContinue onboarding: ${chalk_1.default.cyan(continueUrl)}\n` : ``}
${chalk_1.default.dim('Note: This wizard uses an LLM agent to analyze and modify your project. Please review the changes made.')}

${chalk_1.default.dim(`How did this work for you? Drop me a line: danilo@posthog.com`)}`;
    clack_1.default.outro(outroMessage);
    await analytics_1.analytics.shutdown('success');
}
function buildIntegrationPrompt(context) {
    return `You have access to the PostHog MCP server which provides an integration resource to integrate PostHog into this ${context.framework} project.

Project context:
- Framework: ${context.framework} ${context.version}
- Router: ${context.router}
- TypeScript: ${context.typescript ? 'Yes' : 'No'}
- PostHog API Key: ${context.projectApiKey}
- PostHog Host: ${context.host}

Instructions:
1. Call the PostHog MCP's resource for setup: posthog://integration/workflow/setup/begin
2. Follow all instructions provided

The PostHog MCP will provide specific integration code and instructions. Please follow them carefully. Be sure to look for lockfiles to determine the appropriate package manager to use when installing PostHog. Do not manually edit the package.json file.
`;
}
