// @voiceflow/dtos-interact/build/cjs/trace/visual.dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VisualTraceDTO = exports.VisualTraceImageDTO = void 0;
const zod_1 = require("zod");
const canvas_visibility_enum_1 = require("../enum/canvas-visibility-enum");
const visual_type_enum_1 = require("../enum/visual-type.enum");
const trace_type_enum_1 = require("./trace-type.enum");
const utils_dto_1 = require("./utils.dto");
exports.VisualTraceImageDTO = zod_1.z.object({
    visualType: zod_1.z.literal(visual_type_enum_1.VisualType.IMAGE).optional(),
    image: zod_1.z.string().nullable(),
    device: zod_1.z.string().nullable().optional(),
    options: zod_1.z.object({ loop: zod_1.z.boolean().optional() }).optional(),
    frameType: zod_1.z.string().optional(),
    dimensions: zod_1.z.object({ width: zod_1.z.number(), height: zod_1.z.number() }).nullable().optional(),
    canvasVisibility: zod_1.z.nativeEnum(canvas_visibility_enum_1.CanvasVisibility).optional(),
});
exports.VisualTraceDTO = utils_dto_1.BaseTraceDTO.extend({
    type: zod_1.z.literal(trace_type_enum_1.TraceType.VISUAL),
    /**
     * APL is deprecated, so we only support Image for now.
     */
    payload: exports.VisualTraceImageDTO,
});
