// @posthog/wizard/dist/src/steps/add-mcp-server-to-clients/defaults.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultServerConfig = exports.ALL_FEATURE_VALUES = exports.AVAILABLE_FEATURES = exports.DefaultMCPClientConfig = void 0;
const zod_1 = __importDefault(require("zod"));
exports.DefaultMCPClientConfig = zod_1.default
    .object({
    mcpServers: zod_1.default.record(zod_1.default.string(), zod_1.default.object({
        command: zod_1.default.string().optional(),
        args: zod_1.default.array(zod_1.default.string()).optional(),
        env: zod_1.default.record(zod_1.default.string(), zod_1.default.string()).optional(),
    })),
})
    .passthrough();
exports.AVAILABLE_FEATURES = {
    'Data & Analytics': [
        {
            value: 'dashboards',
            label: 'Dashboards',
            hint: 'Dashboard creation and management',
        },
        {
            value: 'insights',
            label: 'Insights',
            hint: 'Analytics insights and SQL queries',
        },
        {
            value: 'experiments',
            label: 'Experiments',
            hint: 'A/B testing experiments',
        },
        {
            value: 'llm-analytics',
            label: 'LLM Analytics',
            hint: 'LLM usage and cost tracking',
        },
    ],
    'Development Tools': [
        {
            value: 'error-tracking',
            label: 'Error Tracking',
            hint: 'Error monitoring and debugging',
        },
        { value: 'flags', label: 'Feature Flags', hint: 'Feature flag management' },
    ],
    'Platform & Management': [
        {
            value: 'workspace',
            label: 'Workspace',
            hint: 'Organization and project management',
        },
        {
            value: 'docs',
            label: 'Documentation',
            hint: 'PostHog documentation search',
        },
    ],
};
exports.ALL_FEATURE_VALUES = Object.values(exports.AVAILABLE_FEATURES)
    .flat()
    .map((feature) => feature.value);
const getDefaultServerConfig = (apiKey, type, selectedFeatures, local) => {
    const host = local ? 'http://localhost:8787' : 'https://mcp.posthog.com';
    const baseUrl = `${host}/${type === 'sse' ? 'sse' : 'mcp'}`;
    const isAllFeaturesSelected = selectedFeatures &&
        selectedFeatures.length === exports.ALL_FEATURE_VALUES.length &&
        exports.ALL_FEATURE_VALUES.every((feature) => selectedFeatures.includes(feature));
    const urlWithFeatures = selectedFeatures && selectedFeatures.length > 0 && !isAllFeaturesSelected
        ? `${baseUrl}?features=${selectedFeatures.join(',')}`
        : baseUrl;
    return {
        command: 'npx',
        args: [
            '-y',
            'mcp-remote@latest',
            urlWithFeatures,
            '--header',
            `Authorization:\${POSTHOG_AUTH_HEADER}`,
        ],
        env: {
            POSTHOG_AUTH_HEADER: `Bearer ${apiKey}`,
        },
    };
};
exports.getDefaultServerConfig = getDefaultServerConfig;
